File: python/mlmorph/analyser.py
#!/usr/bin/env python

"""
Simple python interface for mlmorph using sfst
"""

from importlib import resources
from typing import List, Tuple

import regex
import sfst

from .foreign_word_detector import check_foreign_word
from .normalizer import normalize


class Analyser:

    ANALYSER_REGEX = regex.compile(r"((?P<root>([^<])+)(?P<tags>(<[^>]+>)+))+")
    POS_REGEX = regex.compile(r"(<(?P<tag>([^>]+))>)+")
    RESOURCE_PATH = "data/malayalam.a"

    def __init__(self):
        """Construct Mlmorph Analyser"""
        self.fsa: str = None
        if resources.files('mlmorph').joinpath(Analyser.RESOURCE_PATH).is_file():
            self.fsa: str = str(resources.files('mlmorph').joinpath(Analyser.RESOURCE_PATH))
        if not self.fsa:
            raise ValueError("Could not read the fsa.")
        sfst.init(self.fsa)


    def analyse(
        self, word: str, weighted: bool = True, foreign_word_check: bool = True
    ) -> List[Tuple[str, int]]:
        """
        Perform a simple morphological analysis lookup.

        Parameters
        ----------

        - word : The word to analyse
        - weighted : Return result with analyser weights. Default: True,
        - foreign_word_check: Use foreign word detector. Default: True

        Returns
        -------
        Array of tuples with analysis string and weight
        """
        word = normalize(word)
        analysis_results = sfst.analyse(word)
        if not len(analysis_results):
            if foreign_word_check and check_foreign_word(word):
                if weighted:
                    analysis_results = [word + "<fw>"]
                else:
                    analysis_results = [word + "<fw>"]

        if not weighted:
            return analysis_results

        processed_result = []
        for aindex in range(len(analysis_results)):
            weight = 0
            analysis = analysis_results[aindex]
            if weighted:
                parsed_result = Analyser.parse_analysis(analysis)
                weight = parsed_result["weight"]
            processed_result.append((analysis, weight))

        return sorted(processed_result, key=lambda tup: tup[1])

    @staticmethod
    def parse_analysis(analysis: str) -> dict:
        """
        Parse the analysis string and return a dict with structured data.

        Parameters
        ----------

        - analysis : The analysis string

        Returns
        -------
        Dictionary with the following structure
        ```
        {
            morphemes: [
                {root:"rootword1", pos:[pos1, pos2] }
                {root:"rootword2", pos:[pos1, pos2] }
            ]
            weight: 200
        }
        ```
        """
        result = {}
        if analysis is None:
            return result

        if analysis[0] == "<":
            analysis = " " + analysis
        match = Analyser.ANALYSER_REGEX.match(analysis)
        if not match:
            raise ValueError("Could not parse the analysis." + analysis)
        roots = match.captures("root")
        morphemes = []
        for rindex in range(len(roots)):
            morpheme = {}
            morpheme["root"] = roots[rindex]
            tags = match.captures("tags")[rindex]
            morpheme["pos"] = Analyser.POS_REGEX.match(tags).captures("tag")
            morphemes.append(morpheme)

        result["morphemes"] = morphemes
        result["weight"] = Analyser.get_weight(morphemes)
        return result

    @staticmethod
    def get_weight(analysis: str) -> int:
        """Analysis with less weight is the best analysis."""
        morpheme_length = len(analysis)
        weight = morpheme_length * 100
        for i in range(morpheme_length):
            pos = analysis[i]["pos"]
            root = analysis[i]["root"]
            for j in range(len(pos)):
                # In general, favor simplicity
                # Prefer analysis with less number of tags
                # Prefer analysis with small length roots
                weight += (
                    len(pos) * 5 + len(root) * 2 + Analyser.get_pos_weight(pos[j]) * 3
                )
        return weight

    @staticmethod
    def get_pos_weight(pos: str) -> int:
        """
        Get the relative weight of a pos tag.
        Less weight is the preferred pos tag.
        """
        WEIGHTS = {
            # Prefer verbs than nouns
            "v": 1,
            "n": 2,
            # Among three letter codes, prefer adv. Then adj, Then pronoun
            "adv": 3,
            "adj": 4,
            "coordinative": 4,
            "v-n-compound": 4,
            "prn": 5,
            # Favor cvb-adv-part-past മുൻവിനയെച്ചം without using its length
            "past": 4,
            "cvb-adv-part-past": 5,
            # Proper noun has high cost
            "np": 5,
            "fw": 1,
        }
        # Use the WEIGHTS or fallback to length
        return WEIGHTS.get(pos, len(pos))

File: rust/src/analyser.rs
use crate::{
    AnalysisResult, Morpheme, ParsedAnalysis, foreign_word_detector::check_foreign_word,
    normalizer::normalize,
};
use regex::Regex;
use sfst::Sfst;
use std::collections::HashMap;

lazy_static::lazy_static! {
    static ref ANALYSER_REGEX: Regex = {
        Regex::new(r"((?P<root>([^<])+)(?P<tags>(<[^>]+>)+))+").unwrap()
    };

    static ref POS_REGEX: Regex = {
        Regex::new(r"(<(?P<tag>([^>]+))>)+").unwrap()
    };
}

pub struct Analyser {
    sfst: Sfst,
}

impl Analyser {
    pub fn new() -> Result<Self, Box<dyn std::error::Error>> {
        let sfst = crate::create_sfst()?;
        Ok(Analyser { sfst })
    }

    pub fn analyse(
        &self,
        word: &str,
        weighted: bool,
        foreign_word_check: bool,
    ) -> Result<Vec<AnalysisResult>, Box<dyn std::error::Error>> {
        let word = normalize(word);
        let mut analysis_results = self.sfst.analyse(&word)?;

        if analysis_results.is_empty() {
            if foreign_word_check && check_foreign_word(&word) == 1 {
                analysis_results = vec![format!("{}<fw>", word)];
            }
        }

        if !weighted {
            return Ok(analysis_results.into_iter().map(|a| (a, 0)).collect());
        }

        let mut processed_result = Vec::new();
        for analysis in analysis_results {
            let weight = if weighted {
                Self::parse_analysis(&analysis)?.weight
            } else {
                0
            };
            processed_result.push((analysis, weight));
        }

        processed_result.sort_by_key(|(_, weight)| *weight);
        Ok(processed_result)
    }

    pub fn parse_analysis(analysis: &str) -> Result<ParsedAnalysis, Box<dyn std::error::Error>> {
        if analysis.is_empty() {
            return Ok(ParsedAnalysis {
                morphemes: Vec::new(),
                weight: 0,
            });
        }

        let analysis = if analysis.starts_with('<') {
            format!(" {}", analysis)
        } else {
            analysis.to_string()
        };

        let _caps = ANALYSER_REGEX
            .captures(&analysis)
            .ok_or("Could not parse the analysis")?;

        let mut morphemes = Vec::new();
        let roots: Vec<&str> = ANALYSER_REGEX
            .captures_iter(&analysis)
            .filter_map(|cap| cap.name("root"))
            .map(|m| m.as_str())
            .collect();

        let tags_list: Vec<&str> = ANALYSER_REGEX
            .captures_iter(&analysis)
            .filter_map(|cap| cap.name("tags"))
            .map(|m| m.as_str())
            .collect();

        for (i, root) in roots.iter().enumerate() {
            if let Some(tags) = tags_list.get(i) {
                let pos: Vec<String> = POS_REGEX
                    .captures_iter(tags)
                    .filter_map(|cap| cap.name("tag"))
                    .map(|m| m.as_str().to_string())
                    .collect();

                morphemes.push(Morpheme {
                    root: root.to_string(),
                    pos,
                });
            }
        }

        let weight = Self::get_weight(&morphemes);
        Ok(ParsedAnalysis { morphemes, weight })
    }

    fn get_weight(morphemes: &[Morpheme]) -> i32 {
        let morpheme_length = morphemes.len() as i32;
        let mut weight = morpheme_length * 100;

        for morpheme in morphemes {
            for pos in &morpheme.pos {
                weight += morpheme.pos.len() as i32 * 5
                    + morpheme.root.len() as i32 * 2
                    + Self::get_pos_weight(pos) * 3;
            }
        }

        weight
    }

    fn get_pos_weight(pos: &str) -> i32 {
        let weights: HashMap<&str, i32> = [
            ("v", 1),
            ("n", 2),
            ("adv", 3),
            ("adj", 4),
            ("coordinative", 4),
            ("v-n-compound", 4),
            ("prn", 5),
            ("past", 4),
            ("cvb-adv-part-past", 5),
            ("np", 5),
            ("fw", 1),
        ]
        .iter()
        .cloned()
        .collect();

        weights.get(pos).copied().unwrap_or(pos.len() as i32)
    }
}
